### DISCUSSION TITLE
Anatolii's Review

### DISCUSSION THREAD

**Discussion by redrose77** [boris-kz/CogAlg#78] [General]

**Comment by redrose77** [comment]
Hi, Chris and Boris.
I am Anatolii, who was introduced by Chris.
I want to thank you for giving me this opportunity for working with your team.
I am very pleased to express my gratitude for what I actually want to know.
I have found your idea is very interesting and innovative to me.

ANN has been showing its wonderful abilities in many fields, especially image recognition, classification and object detection field. ANN is mainly based on supervised learning rather than unsupervised learning in solving practical problems.
Since studying ANN, I have been thinking of ANN's shortcoming(comparison to ANN and BNN), but I haven't found it clearly.
After I have read your Readme, I had some understandings about it in some extents.
I have studied and implemented ANN for several years, so I have some understandings on supervised learning but few on unsupervised learning.
So I have reviewed for understanding CogAlg's main idea for several days, but feel very complicated in understanding cross-comparison.
Before telling my opinion to you, I'd like to know about "cross-comparison" in details.
Please let me know about "cross-comparison" and your clustering method in details.
Thank you in advance.

**Comment by boris-kz** [comment]
Hi Anatolii, appreciate your interest. 

Cross-comparison is comparison of each input to all other inputs within a certain distance. Maximal comparison distance and input composition are incremental with elevation: greater on higher levels of the algorithm. That means they should be defined bottom-up, starting with nearest-neighbor pixel inputs in 1D, please see cross_comp() in line_patterns: https://github.com/boris-kz/CogAlg/blob/master/line_1D_alg/line_patterns.py. 

Connectivity clustering here is primarily by the sign of match deviation, which is a product of cross-comp. Initially it's the same as image segmentation, see form_P_ in line_patterns, or form_blob in frame_blobs. Subsequent internal and external recursion will form hierarchical clusters: https://en.wikipedia.org/wiki/Hierarchical_clustering.

So, I think the hard part here is the idea of incremental complexity: there is no fixed input format or processing, only recursive meta-process of incrementing both on each higher level of search. Which can only be defined bottom-up, starting with line_patterns.

**Reply by redrose77** [reply]
Thanks for your kind explanation, Boris.
Very pleased to pick up new understanding of CogAlgorithm.
I am thinking how much I can contribute to developing CogAlgorithm with my knowledge and creativity.
Best Regards to you.

**Reply by boris-kz** [reply]
Thanks! Knowledge and creativity is great, but main thing in this project is consistency: everything is supposed to be designed to maximize predictive value (projected match), from the bottom up.

**Reply by redrose77** [reply]
Thanks for your kind advice, Boris.
I think you are very skillful in this field.
Can't wait to work with you.
I will try to understand CogAlgorithm as soon as possible.
I hope you to help me in the future.
Sincerely, Anatolli.

**Reply by boris-kz** [reply]
It's not a field, just one project :). I don't know of any other theory-first work, ML is a heavily experimental field.
